src/Compilers/Core/Portable/InternalUtilities/IncrementalHash.cs

```diff
@@ -0,0 +1,99 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Reflection;
+using Microsoft.CodeAnalysis;
+
+namespace Roslyn.Utilities
+{
+    internal abstract class IncrementalHash : IDisposable
+    {
+        public void AppendData(byte[] data) => AppendData(data, 0, data.Length);
+
+        public abstract void AppendData(byte[] data, int offset, int count);
+        public abstract byte[] GetHashAndReset();
+        public abstract void Dispose();
+
+        public static IncrementalHash Create(AssemblyHashAlgorithm hashAlgorithm)
+        {
+            if (PortableShim.IncrementalHash.Type != null)
+            {
+                return new Core(hashAlgorithm);
+            }
+            else
+            {
+                return new Desktop(hashAlgorithm);
+            }
+        }
+
+        private sealed class Core : IncrementalHash
+        {
+            // IncrementalHash
+            private readonly IDisposable _incrementalHashImpl;
+
+            internal Core(AssemblyHashAlgorithm hashAlgorithm)
+            {
+                var name = CryptographicHashProvider.TryGetHashAlgorithmNameObj(hashAlgorithm);
+                _incrementalHashImpl = PortableShim.IncrementalHash.CreateHash(name);
+            }
+
+            public override void AppendData(byte[] data, int offset, int count) => PortableShim.IncrementalHash.AppendData(_incrementalHashImpl, data, offset, count);
+            public override byte[] GetHashAndReset() => PortableShim.IncrementalHash.GetHashAndReset(_incrementalHashImpl);
+            public override void Dispose() => _incrementalHashImpl.Dispose();
+        }
+
+        private sealed class Desktop : IncrementalHash
+        {
+            // HashAlgorithm
+            private readonly IDisposable _hashAlgorithmImpl;
+
+            private ArraySegment<byte> _pending;
+
+            internal Desktop(AssemblyHashAlgorithm hashAlgorithm)
+            {
+                _hashAlgorithmImpl = CryptographicHashProvider.TryGetAlgorithmImpl(hashAlgorithm);
+            }
+
+            public override void AppendData(byte[] data, int offset, int count)
+            {
+                if (_pending.Array != null)
+                {
+                    TransformBlock(_pending.Array, _pending.Offset, _pending.Count);
+                }
+
+                _pending = new ArraySegment<byte>(data, offset, count);
```

Comment:
This doesn't feel safe, consumer might assume that the data has been processed and might change content of the array.